### DISCUSSION TITLE
Jingles playlist does not play when there are listener requests in the queue

### DISCUSSION THREAD

**Discussion by BornKillaz** [AzuraCast/AzuraCast#7325] [Q&A] [answered]
I've come across an issue where the jingles playlist stops playing when listener requests fill the queue. 

Has anyone else experienced this problem, or is this expected behavior?

I have my jingles playlist set to play every 2 songs.

**Comment by BornKillaz** [comment]
So my listener requests are made via AzuraCast API, if the issue I describe above is the normal functionality of AzuraCast, probably the solution is to programmatically set up Jingles between requests? 

Don't know if that would work either...

**Reply by gAlleb** [reply]
Hi, it's not an issue but an expected behavior.

What is `requests queue`?  Here is the code form default LS configuration of Azuracast:
<img width="600" alt="image" src="https://github.com/user-attachments/assets/6e7d1cee-89e8-4f00-95aa-0b6e527167ab">

`request.queue` acts like a `fallback` to your normal playback (including all of your playlists, jingles included) thus any jingles or other playlist or programs you have are interrupted by `request.queue` when it's active. 

I think that is an intended functionality and I thinks it's implemented in the API that way. 

If you want to play jingles between requests you have to make sure that jingles will interrupt `request.queue`. This can be achieved with Advanced playlists and editing of the 3rd box of LS config:

```
radio = rotate(weights=[1, 4], [playlist_jingles, radio])
```

This way `rotate` mechanism will be applied to the source after `request.queue` fallback thus interrupting it in order to play jingles.

**Reply by BornKillaz** [reply]
Hi! 

First of all, thank you so much for the time you took to help me out on this... However, I've tried as you said and unfortunately it also didn't work, jingles wont play when there are  queue requests.

I've selected advanced playlist, put the correct jingles playlist name and parameters in the 3rd box in the liquidsoap configuration, in my case: `radio = rotate(weights=[1, 2], [playlist_sweepers, radio])` but it didn't work...

**Reply by gAlleb** [reply]
Indeed, strange. And have you restarted broadcasting? (just to be sure). Also worth mentioning that you won't see `jingles` in upcoming song queue because they will not be controlled by API but you should hear them.

**Reply by BornKillaz** [reply]
Yes, I did restart (reload in fact). Is that ok? Or it must be a full station restart?
And yes, I noticed they wont show up in the  song queue, and I make sure I let a few requests pass by just to be sure. And nothing. The jingles playlist should play every 2 other songs, but no luck... 

They played fine when there were no requests, but as soon as the requests kick in, it wont play anymore between those requested tracks.

**Reply by gAlleb** [reply]
That shouldn't be the case. Very strange.

`playlist_sweepers = playlist(id="playlist_sweepers",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_sweepers.m3u")` should be in your LS config and `radio = rotate(weights=[1, 2], [playlist_sweepers, radio])` in the 3rd box. And restart broadcasting.

Azuracast requests can't control advanced playlists configuration. This goes against the LS script. Hm...

**Reply by BornKillaz** [reply]
OK, now I've got it working! Thanks a lot, gAlleb. So far, it looks good, except that the files don't seem to be parsed by AutoCue. They play without respecting cue points or pre-tagged replay gain. Additionally, since I have sound plugins, those WAV sweepers distort a lot. I don't know what to do... Do you have any idea?

**Reply by gAlleb** [reply]
All files are processed, because when you create a playlist every track inside of it gets annotated with `autocue:annotate:`. so as not to be unsubstantiated I've just tested it and it works fine with all tracks being autocued. What do your logs say when LS pushes jingle track?

**Reply by gAlleb** [reply]
Also make sure to upgrade Azuracast due to 04c6c857fd786aa4a19ed9b921b9b37ab7b5bc24 . This commit moves `amplify` operator below the 3rd box so your Advanced playlists could be normalized.

**Reply by BornKillaz** [reply]
I'm on Rolling Release 2ff8c76, it's the latest one.

And I have  `radio = rotate(weights=[1, 2], [playlist_sweepers, radio])` on the 3rd box.

These are the liquidsoap logs:
```
2024/07/15 15:56:11 [decoder.ffmpeg:3] Requested content-type for "/var/azuracast/stations/basstik_radio/media/BASSTIK/SWEEPERS/basstik_radio_-_non-stop_bangers,_all_day,_every_day_(st).wav": {audio=pcm(stereo)}
2024/07/15 15:56:11 [decoder.ffmpeg:3] FFmpeg recognizes "/var/azuracast/stations/basstik_radio/media/BASSTIK/SWEEPERS/basstik_radio_-_non-stop_bangers,_all_day,_every_day_(st).wav" as audio: {codec: pcm_s16le, 48000Hz, 2 channel(s)}, video: {codec: png, 600x600, rgba}
2024/07/15 15:56:11 [decoder.ffmpeg:3] Decoded content-type for "/var/azuracast/stations/basstik_radio/media/BASSTIK/SWEEPERS/basstik_radio_-_non-stop_bangers,_all_day,_every_day_(st).wav": {audio=pcm(stereo)}
2024/07/15 15:56:11 [lang:3] API feedback - Sending POST request to 'http://127.0.0.1:6010/api/internal/2/liquidsoap/feedback' with body: {}
2024/07/15 15:56:11 [autocue.cue_file:3] Now autocueing: "/var/azuracast/stations/basstik_radio/media/BASSTIK/SWEEPERS/basstik_radio_-_non-stop_bangers,_all_day,_every_day_(st).wav"
2024/07/15 15:56:11 [autocue.cue_file:3] Blank (silence) skipping active: false, set to 0. s
2024/07/15 15:56:11 [autocue.cue_file:3] Clipping prevention active: false
2024/07/15 15:56:11 [autocue.cue_file:3] Writing tags: false, including ReplayGain: false
2024/07/15 15:56:11 [autocue.cue_file:3] Writing metadata to /tmp/cue_file679b6e.json
2024/07/15 15:56:11 [lang:3] API feedback - Response (200): false
2024/07/15 15:56:11 [autocue.cue_file:3] cue_file result for "/var/azuracast/stations/basstik_radio/media/BASSTIK/SWEEPERS/basstik_radio_-_non-stop_bangers,_all_day,_every_day_(st).wav": {"duration": 7.262854, "liq_cue_duration": 7.1, "liq_cue_in": 0.0, "liq_cue_out": 7.1, "liq_cross_start_next": 7.1, "liq_longtail": false, "liq_sustained_ending": true, "liq_loudness": "-7.13 LUFS", "liq_loudness_range": "3.16 LU", "liq_amplify": "-10.87 dB", "liq_amplify_adjustment": "0.00 dB", "liq_reference_loudness": "-18.00 LUFS", "liq_blankskip": 0.0, "liq_blank_skipped": false, "liq_true_peak": 0.95, "liq_true_peak_db": "-0.45 dBFS"}
2024/07/15 15:56:11 [autocue.cue_file:3] Can't override liq_amplify from replaygain_track_gain, replaygain_reference_loudness missing.
2024/07/15 15:56:11 [autocue.cue_file:2] Given fade-out duration (0.1 s) exceeds available time, using 0. s.
2024/07/15 15:56:11 [autocue.cue_file:3] Metadata added/corrected for "/var/azuracast/stations/basstik_radio/media/BASSTIK/SWEEPERS/basstik_radio_-_non-stop_bangers,_all_day,_every_day_(st).wav":
2024/07/15 15:56:11 [autocue.cue_file:3] ("duration", "7.262854")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_amplify", "-10.87 dB")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_amplify_adjustment", "0.00 dB")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_autocue", "cue_file")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_blank_skipped", "false")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_blankskip", "0.00")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_cross_duration", "2.50")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_cross_start_next", "7.1")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_cue_duration", "7.10")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_cue_in", "0.00")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_cue_out", "7.1")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_fade_in", "0.1")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_fade_in_delay", "0.50")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_fade_out", "0.")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_fade_out_delay", "2.50")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_longtail", "false")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_loudness", "-7.13 LUFS")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_loudness_range", "3.16 LU")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_reference_loudness", "-18.00 LUFS")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_sustained_ending", "true")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_true_peak", "0.95")
2024/07/15 15:56:11 [autocue.cue_file:3] ("liq_true_peak_db", "-0.45 dBFS")
2024/07/15 15:56:11 [autocue.cue_file:3] ("replaygain_track_gain", "-10.88 dB")
[wav @ 0x7f9d62a4c000] Discarding ID3 tags because more suitable tags were found.
[wav @ 0x7f9d62a4c000] Discarding ID3 tags because more suitable tags were found.
2024/07/15 15:56:11 [decoder.ffmpeg:3] Requested content-type for "/var/azuracast/stations/basstik_radio/media/BASSTIK/SWEEPERS/basstik_radio_-_non-stop_bangers,_all_day,_every_day_(st).wav": {audio=pcm(stereo)}
2024/07/15 15:56:11 [decoder.ffmpeg:3] FFmpeg recognizes "/var/azuracast/stations/basstik_radio/media/BASSTIK/SWEEPERS/basstik_radio_-_non-stop_bangers,_all_day,_every_day_(st).wav" as audio: {codec: pcm_s16le, 48000Hz, 2 channel(s)}, video: {codec: png, 600x600, rgba}
2024/07/15 15:56:11 [decoder.ffmpeg:3] Decoded content-type for "/var/azuracast/stations/basstik_radio/media/BASSTIK/SWEEPERS/basstik_radio_-_non-stop_bangers,_all_day,_every_day_(st).wav": {audio=pcm(stereo)}
2024/07/15 15:56:14 [decoder:2] Decoding "/var/azuracast/stations/basstik_radio/media/BASSTIK/SWEEPERS/basstik_radio_-_the_sound_of_now,_the_beat_of_wow_(st).wav" ended: Ffmpeg_decoder.End_of_file.
[wav @ 0x7f9d640bd500] Discarding ID3 tags because more suitable tags were found.
[wav @ 0x7f9d640bd500] Discarding ID3 tags because more suitable tags were found.
2024/07/15 15:56:15 [playlist_sweepers:3] Prepared "/var/azuracast/stations/basstik_radio/media/BASSTIK/SWEEPERS/basstik_radio_-_keep_listening,_its_basstik_radio_(st).wav" (RID 64).
2024/07/15 15:56:15 [switch.7:3] Switch to interrupting_fallback with forgetful transition.
2024/07/15 15:56:15 [replay_metadata.70:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [show_meta:3] ("duration", "193.01")
2024/07/15 15:56:15 [show_meta:3] ("liq_amplify", "-9.87 dB")
2024/07/15 15:56:15 [show_meta:3] ("liq_amplify_adjustment", "0.00 dB")
2024/07/15 15:56:15 [show_meta:3] ("liq_autocue", "cue_file")
2024/07/15 15:56:15 [show_meta:3] ("liq_blank_skipped", "false")
2024/07/15 15:56:15 [show_meta:3] ("liq_blankskip", "5.00")
2024/07/15 15:56:15 [show_meta:3] ("liq_cross_duration", "2.50")
2024/07/15 15:56:15 [show_meta:3] ("liq_cross_start_next", "186.75")
2024/07/15 15:56:15 [show_meta:3] ("liq_cue_duration", "189.05")
2024/07/15 15:56:15 [show_meta:3] ("liq_cue_in", "0.00")
2024/07/15 15:56:15 [show_meta:3] ("liq_cue_out", "189.05")
2024/07/15 15:56:15 [show_meta:3] ("liq_fade_in", "0.10")
2024/07/15 15:56:15 [show_meta:3] ("liq_fade_in_delay", "1.80")
2024/07/15 15:56:15 [show_meta:3] ("liq_fade_out", "2.30")
2024/07/15 15:56:15 [show_meta:3] ("liq_fade_out_delay", "0.20")
2024/07/15 15:56:15 [show_meta:3] ("liq_longtail", "false")
2024/07/15 15:56:15 [show_meta:3] ("liq_loudness", "-8.13 LUFS")
2024/07/15 15:56:15 [show_meta:3] ("liq_loudness_range", "7.16 LU")
2024/07/15 15:56:15 [show_meta:3] ("liq_reference_loudness", "-18.00 LUFS")
2024/07/15 15:56:15 [show_meta:3] ("liq_sustained_ending", "false")
2024/07/15 15:56:15 [show_meta:3] ("liq_true_peak", "1.11")
2024/07/15 15:56:15 [show_meta:3] ("liq_true_peak_db", "0.88 dBFS")
2024/07/15 15:56:15 [show_meta:3] ("media_id", "6007")
2024/07/15 15:56:15 [show_meta:3] ("replaygain_track_gain", "-9.88 dB")
2024/07/15 15:56:15 [show_meta:3] Now playing: Rowdy Rebel, Fetty Luciano & Fivio Foreign - Posture (Feat. Fetty Luciano & Fivio Foreign)
2024/07/15 15:56:15 [lang:3] API nextsong - Sending POST request to 'http://127.0.0.1:6010/api/internal/2/liquidsoap/nextsong' with body: 
2024/07/15 15:56:15 [cross:3] Analysis: -11.925590dB / -35.183783dB (2.46s / 2.46s)
2024/07/15 15:56:15 [cross:3] Computing crossfade transition over reconciled duration with 2.46s ending and 2.46s starting buffered data.
2024/07/15 15:56:15 [source.98:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [audio.add.31:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [fade_in.31:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [fade_scale.31:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [track_amplify.62:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [blank.33:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [on_metadata.61:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [on_track.63:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [cross_after.31:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [buffer.69:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [fade_out.31:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [track_amplify.63:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [on_frame.63:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [on_metadata.62:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [on_track.64:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [cross_before.31:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [buffer.68:3] Content type is {audio=pcm(stereo)}.
2024/07/15 15:56:15 [lang:3] API nextsong - Response (200): autocue:annotate:title="Bubbly",artist="Conki, K!llx & Naja",duration="123.02",song_id="d3158231c3977282b04224679588ec30",media_id="5750",liq_amplify="-11.15 dB",liq_amplify_adjustment="0.00 dB",liq_autocue="cue_file",liq_blank_skipped="false",liq_blankskip="5.00",liq_cross_duration="2.50",liq_cross_start_next="120.25",liq_cue_duration="122.05",liq_cue_in="0.00",liq_cue_out="122.05",liq_fade_in="0.10",liq_fade_in_delay="0.01",liq_fade_out="1.80",liq_fade_out_delay="0.70",liq_longtail="false",liq_loudness="-6.85 LUFS",liq_loudness_range="3.45 LU",liq_reference_loudness="-18.00 LUFS",liq_sustained_ending="false",liq_true_peak="1.21",liq_true_peak_db="1.62 dBFS",replaygain_track_gain="-11.16 dB",replaygain_track_peak="1.21",playlist_id="58":media:POP/conki,_kllx__naja_-_bubbly.m4a
```

On normal playlist mode, these sweepers play without any sound problems and cue points are respected. Don't know why is not the case in here.

**Reply by gAlleb** [reply]
If you are pasting logs please embrace them in ` ``` ` at the beginning and the end because it’s impossible to read them now.

**Reply by BornKillaz** [reply]
> If you are pasting logs please embrace them in `` ``` `` at the beginning and the end because it’s impossible to read them now.

Thank you. Done.

**Reply by gAlleb** [reply]
Hard to say what is happening. Do logs show `[show_meta]` for sweepers? Cause it’s hard to say from this small chunk. Right now I can’t think of a reason why it’s happening, because autocue computes those wav files.

**Reply by BornKillaz** [reply]
Logs show [show_meta:3]  for these sweepers. I'll try with mp3 version of these files, maybe the wav format is the cause of it, don't really know why, becouse, as I said, they run perfectly fine when set up with Azuracast.

**Reply by gAlleb** [reply]
So I have tested with different `wav` and `mp3`s and so on. Everything seems to be good and works as expected.

Below you can see LUFS target of the stream -13.2 LUFS because I use

```
radio = amplify(1., override="liq_amplify", radio)
radio = amplify(lin_of_dB(5.0), override=null(), radio)
```
instead of default amplify. I takes default -18 by autocue and plus 5.

<img width="1071" alt="Снимок экрана 2024-07-15 в 19 41 38" src="https://github.com/user-attachments/assets/a5cbee97-39f5-4f97-8739-897a4b927210">

<img width="1036" alt="image" src="https://github.com/user-attachments/assets/69d348d0-4341-49b8-a00f-f0209fbb4e7d">

**Comment by BornKillaz** [comment]
Ok, I've just tried with mp3 files, same behavior.

Liquidsoap Logs (first comes the .wav,'s, then at the end the mp3's):
https://drive.google.com/file/d/18UEUZVYE0EQG1MfCsF4NK1Prj4A8HRr2/view

Replay gain is not taken into consideration, nor cue points.

I have this set in AzuraCast in Jingle mode, selected advanced playlist, is a folder based playlist and i have it put it in the 3rd liquidsoap box as mentioned. Cannot figure what it's going on, so I'm forced to move back to the playlist rotation manage by AzuraCast until I can sort this out.

**Reply by gAlleb** [reply]
Now everything makes sense to me. You set `Jingle mode` and this results in `playlist = drop_metadata(playlist)` and you drop it))

So uncheck `Jingle mode`. When using advanced configuration we shouldn't drop metadata. Instead as a workaround we can use `jingle_mode=true`. Hopefully this would exclude sweepers from showing in nowplaying. But at least now it's clear what went wrong. `playlist = drop_metadata(playlist)` (Jingle mode switch) removes all metadata for `autocue`.

Since we have annotations already in our playlists - `prefix="annotate:jingle_mode=true:",` won't work. The easiest way then - just to add metadata `jingle_mode=true` into the files (jingles, sweepers).

So:

1. Edit your files. Add tag `jingle_mode=true`.
2. Uncheck `Jingle mode` in Playlist Configuration.
3. Add `radio = rotate(weights=[1, 2], [playlist_sweepers, radio])` into 3rd box.

It's a good practice to add `jingle_mode=true` into your file tags. E.g. All of my jingles/sweepers/ads have it. However not sure if it would work) `prefix="annotate:jingle_mode=true:",`  has worked before, but now since `autocue` has been introduced we can't annotate this way anymore. So it may be a bummer after all.

**Comment by gAlleb** [comment, marked as answer]
Ok, last but not least. After some testing here is the solution:

1. Create `new` Advanced playlist and `check` `Jingle mode` (yes) toggler! (I suggest you create new playlist - this way everything will work from the start) And save it.
This would result in 
```
playlist_sweepers = playlist(id="playlist_sweepers",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_sweepers.m3u")

playlist_sweepers = drop_metadata(playlist_sweepers)

ignore(playlist_sweepers)
```
in your config. And will also write `jingle_mode=true` into this playlist. Also add files into this playlist.

2. Go to 3rd box and redefine playlist again. Just copy and paste `playlist_sweepers = playlist(id="playlist_sweepers",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_sweepers.m3u")` this line. Don't copy `drop_metadata` line and `ignore` line.
3.  Add rotate:
```
radio = rotate(weights=[1, 2], [playlist_sweepers, radio])
```
4. In the end this is what you will see in the 3rd box:
```
playlist_sweepers = playlist(id="playlist_sweepers",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_sweepers.m3u")
radio = rotate(weights=[1, 2], [playlist_sweepers, radio])
```

5. Restart and everything should work.

**Reply by BornKillaz** [reply]
Can't thank you enough for the time you've taken to analyze this and provide solutions.

I will try it soon and let you know how it goes.

**Reply by BornKillaz** [reply]
> Ok, last but not least. After some testing here is the solution:
> 
> 1. Create `new` Advanced playlist and `check` `Jingle mode` (yes) toggler! (I suggest you create new playlist - this way everything will work from the start) And save it.
>    This would result in
> 
> ```
> playlist_sweepers = playlist(id="playlist_sweepers",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_sweepers.m3u")
> 
> playlist_sweepers = drop_metadata(playlist_sweepers)
> 
> ignore(playlist_sweepers)
> ```
> 
> in your config. And will also write `jingle_mode=true` into this playlist. Also add files into this playlist.
> 
> 2. Go to 3rd box and redefine playlist again. Just copy and paste `playlist_sweepers = playlist(id="playlist_sweepers",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_sweepers.m3u")` this line. Don't copy `drop_metadata` line and `ignore` line.
> 3. Add rotate:
> 
> ```
> radio = rotate(weights=[1, 2], [playlist_sweepers, radio])
> ```
> 
> 4. In the end this is what you will see in the 3rd box:
> 
> ```
> playlist_sweepers = playlist(id="playlist_sweepers",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_sweepers.m3u")
> radio = rotate(weights=[1, 2], [playlist_sweepers, radio])
> ```
> 
> 5. Restart and everything should work.

Works like a charm! 

Now my sweepers are played between requests and also manual file schedules. Meta-data is not passed to the 'now playing' info, and they are listed in the played log in AzuraCast. They also respect cue points and replay gain, so this is perfect now.

Thank you very much @gAlleb 🥇 !
